metagamer/environments/qagent.py

```diff
@@ -0,0 +1,144 @@
+from typing import Tuple, Optional, Any, Dict
+
+GAMMA = 0.9
+import random
+import numpy as np
+from metagamer.environments import tictactoe
+from metagamer.environments import qtable
+
+
+def exploit(epsilon):
+    """exploitation increases as epsilon decreases"""
+    return random.random() > epsilon
+
+
+class TicTacToeTableEnv(tictactoe.TicTacToeEnv):
+    """An TicTacTo subclass environment adapted to Q learning"""
+
+    SYMBOLS = {1: "X", -1: "O", 0: " "}
+
+    def _get_obs(self) -> str:
+        return "".join([self.SYMBOLS[i] for i in self.board.flatten()])
+
+    def action_to_2d(self, action: int):
+        """Turns the integer position (in array of board) into 2D coordinate in form of tuple """
+        new_action = tuple(
+            np.unravel_index(action, (self.board.shape[0], self.board.shape[1]))
+        )
+        return new_action
+
+    def action_to_1d(self, action: Tuple[int, int]) -> int:
+        """Take a 2D coordinate on the board, and turn it into an integer position on the board"""
+        return self.board.shape[1] * action[0] + action[1]
+
+
+class Agent:
```

Comment:
So I've already mentioned this in person, but I'd be more willing to throw away this abstraction - this code was overly specific to the problem (Cartpole) and more focussed on switching in differing qtable implementations.

Lot of this code isn't very good and useful - for example the get_action makes this code almost impossible to extend to 2 agents and instead should be something that each agent handles itself. 
The epsilon decay can be included in trainer as I did in #8